src/nc_connection.h

```diff
@@ -38,14 +38,44 @@ typedef void (*conn_msgq_t)(struct context *, struct conn *, struct msg *);
 typedef void (*conn_post_connect_t)(struct context *ctx, struct conn *, struct server *server);
 typedef void (*conn_swallow_msg_t)(struct conn *, struct msg *, struct msg *);
 
+/*
+ * There are few kinds of connections, client-bound, server-bound, redis,
+ * memcache, proxy, etc. We uniquely identify a connection kind using a single
+ * enum, otherwise the code becomes a tangle of if's.
+ */
+enum conn_kind {
```

Comment:
instead of creating a the all possible combinations of connection (client | proxy | server) and protocol (redis | memcache), I think it would make sense to keep the original design of three types of connection (client | proxy | server) and a separate field for protocol.

The questions then becomes how to you express CONN_SIGNAL. For that I propose we should introduce a new structure (struct signal or something like that) with signatures something like this:

struct conn {
  uint32_t tag;  /* initialized to some conn specific magic let's say 0x00000001 */
}

struct signal {
  uint32_t tag; /* initialized to some unique signal magic let's say 0x0000002 */
}

This way when you in the event callback you can have different handlers for conn based events and signal based events

Also I've been doing cleanups based on your code in the branch "config-reload". Would be nice to incorporate those cleanups. thanks!